src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseInviteService.cs

```diff
@@ -0,0 +1,225 @@
+                email: email);
+            if (emailResult.IsLeft)
+            {
+                return emailResult;
+            }
+
+            await _userInviteRepository.Create(
+                email: email,
+                role: Role.Analyst,
+                createdById: _userService.GetUserId());
+            await _userReleaseInviteRepository.CreateManyIfNotExists(
+                releaseIds: releaseIds,
+                email: email,
+                releaseRole: Contributor,
+                emailSent: true,
+                createdById: _userService.GetUserId());
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Unit>> CreateExistingUserContributorInvite(List<Guid> releaseIds,
+            Guid userId, string email, string publicationTitle)
+        {
+            // check the user doesn't already have the user release roles
+            var existingReleaseRoleReleaseIds = _contentDbContext.UserReleaseRoles
+                .AsQueryable()
+                .Where(urr =>
+                    releaseIds.Contains(urr.ReleaseId)
+                    && urr.Role == Contributor
+                    && urr.UserId == userId)
+                .Select(urr => urr.ReleaseId)
+                .ToList();
+
+            var missingReleaseRoleReleaseIds = releaseIds
+                .Where(releaseId => !existingReleaseRoleReleaseIds.Contains(releaseId))
+                .ToList();
+
+            if (missingReleaseRoleReleaseIds.IsNullOrEmpty())
+            {
+                return ValidationActionResult(UserAlreadyHasReleaseRoles);
+            }
+
+            var emailResult = await SendContributorInviteEmail(
+                publicationTitle: publicationTitle,
+                releaseIds: missingReleaseRoleReleaseIds,
+                email: email);
+            if (emailResult.IsLeft)
+            {
+                return emailResult;
+            }
+
+            await _userReleaseRoleRepository.CreateManyIfNotExists(
+                userId: userId,
+                releaseIds: missingReleaseRoleReleaseIds,
+                role: Contributor,
+                createdById: _userService.GetUserId()
+            );
+
+            await _userReleaseInviteRepository.CreateManyIfNotExists(
+                releaseIds: missingReleaseRoleReleaseIds,
+                email: email,
+                releaseRole: Contributor,
+                emailSent: true,
+                createdById: _userService.GetUserId(),
+                accepted: true);
+
+            return Unit.Instance;
+        }
+
+        private async Task<Either<ActionResult, Publication>> ValidateReleaseIds(Publication publication,
+            List<Guid> releaseIds)
+        {
+            var distinctReleaseIds = releaseIds.Distinct().ToList();
+            if (distinctReleaseIds.Count != releaseIds.Count)
+            {
+                throw new ArgumentException("releaseIds should not contain duplicates");
```

Comment:
Suspect I'm guilty of not doing this but we can use `nameof` so that if the field name is refactored the exception message gets refactored as well. It's also possible to specify an additional param referring to the offending argument name. See the last param here:

`throw new ArgumentException($"{nameof(releaseIds)} should not contain duplicates", nameof(releaseIds));`